{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 0.1 Import "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-03-21T23:04:39.176458Z",
     "start_time": "2020-03-21T23:04:39.171897Z"
    }
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import glob\n",
    "from Bio import SeqIO, Seq, pairwise2, SeqRecord, Align\n",
    "from os.path import join\n",
    "import pickle\n",
    "\n",
    "from tqdm import tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-03-21T23:20:18.785718Z",
     "start_time": "2020-03-21T23:20:02.455919Z"
    }
   },
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Load:\n",
    "BLAST SAM file\n",
    "BLAST alignment \n",
    "- A) Genomic fasta sequences, \n",
    "- B) Reference Genomic Annotation (GCF_009858895.2)\n",
    "- C) Reference Gene Annotation (NC_045512.2)\n",
    "- D) S-protein fasta protein sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-03-22T01:56:15.580306Z",
     "start_time": "2020-03-22T01:56:15.577208Z"
    }
   },
   "outputs": [],
   "source": [
    "cmd = \"python /home/isshamie/software/ebi/emboss_water.py --stype dna --bsequence ../S_gene.fasta --asequence genomeA.fasta --email isshamie@ucsd.edu\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-03-22T00:56:45.275647Z",
     "start_time": "2020-03-22T00:56:45.257942Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "ename": "IndexError",
     "evalue": "list index out of range",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mIndexError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-6-88c060adabda>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0malign\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mAlignIO\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"../data/scrap/emboss_water-R20200322-004539-0954-87715542-p2m.aln\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"clustal\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/software/anaconda2/envs/coevolve/lib/python3.7/site-packages/Bio/AlignIO/__init__.py\u001b[0m in \u001b[0;36mread\u001b[0;34m(handle, format, seq_count, alphabet)\u001b[0m\n\u001b[1;32m    433\u001b[0m     \u001b[0miterator\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mparse\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mhandle\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mformat\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mseq_count\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0malphabet\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    434\u001b[0m     \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 435\u001b[0;31m         \u001b[0mfirst\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnext\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0miterator\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    436\u001b[0m     \u001b[0;32mexcept\u001b[0m \u001b[0mStopIteration\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    437\u001b[0m         \u001b[0mfirst\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/software/anaconda2/envs/coevolve/lib/python3.7/site-packages/Bio/AlignIO/__init__.py\u001b[0m in \u001b[0;36mparse\u001b[0;34m(handle, format, seq_count, alphabet)\u001b[0m\n\u001b[1;32m    380\u001b[0m         \u001b[0;31m# TODO: As of Python 3.3, can write \"yield from i\" instead. See PEP380.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    381\u001b[0m         \u001b[0;31m# For loop imposes some overhead... wait until we drop Python 2.7 to fix it.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 382\u001b[0;31m         \u001b[0;32mfor\u001b[0m \u001b[0ma\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mi\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    383\u001b[0m             \u001b[0;32myield\u001b[0m \u001b[0ma\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    384\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/software/anaconda2/envs/coevolve/lib/python3.7/site-packages/Bio/AlignIO/ClustalIO.py\u001b[0m in \u001b[0;36m__next__\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    111\u001b[0m         \u001b[0;31m# Whitelisted headers we know about\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    112\u001b[0m         \u001b[0mknown_headers\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m'CLUSTAL'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'PROBCONS'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'MUSCLE'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'MSAPROBS'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'Kalign'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 113\u001b[0;31m         \u001b[0;32mif\u001b[0m \u001b[0mline\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstrip\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mknown_headers\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    114\u001b[0m             raise ValueError(\"%s is not a known CLUSTAL header: %s\" %\n\u001b[1;32m    115\u001b[0m                              (line.strip().split()[0],\n",
      "\u001b[0;31mIndexError\u001b[0m: list index out of range"
     ]
    }
   ],
   "source": [
    "align = AlignIO.read(\"../data/scrap/emboss_water-R20200322-004539-0954-87715542-p2m.aln\", \"clustal\",)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-03-22T00:50:41.598757Z",
     "start_time": "2020-03-22T00:50:41.581980Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/isshamie/software/anaconda2/envs/coevolve/lib/python3.7/site-packages/ipykernel_launcher.py:2: DeprecationWarning: 'U' mode is deprecated\n",
      "  \n"
     ]
    },
    {
     "ename": "ValueError",
     "evalue": "######################################## is not a known CLUSTAL header: CLUSTAL, PROBCONS, MUSCLE, MSAPROBS, Kalign",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-2-1343a75c5622>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mBio\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mSeqIO\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mAlignIO\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;32mwith\u001b[0m \u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"../data/scrap/emboss_water-R20200322-004539-0954-87715542-p2m.aln.txt\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"rU\"\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mhandle\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m     \u001b[0;32mfor\u001b[0m \u001b[0mrecord\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mAlignIO\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mparse\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mhandle\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"clustal\"\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      4\u001b[0m         \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrecord\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mid\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/software/anaconda2/envs/coevolve/lib/python3.7/site-packages/Bio/AlignIO/__init__.py\u001b[0m in \u001b[0;36mparse\u001b[0;34m(handle, format, seq_count, alphabet)\u001b[0m\n\u001b[1;32m    380\u001b[0m         \u001b[0;31m# TODO: As of Python 3.3, can write \"yield from i\" instead. See PEP380.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    381\u001b[0m         \u001b[0;31m# For loop imposes some overhead... wait until we drop Python 2.7 to fix it.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 382\u001b[0;31m         \u001b[0;32mfor\u001b[0m \u001b[0ma\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mi\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    383\u001b[0m             \u001b[0;32myield\u001b[0m \u001b[0ma\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    384\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/software/anaconda2/envs/coevolve/lib/python3.7/site-packages/Bio/AlignIO/ClustalIO.py\u001b[0m in \u001b[0;36m__next__\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    114\u001b[0m             raise ValueError(\"%s is not a known CLUSTAL header: %s\" %\n\u001b[1;32m    115\u001b[0m                              (line.strip().split()[0],\n\u001b[0;32m--> 116\u001b[0;31m                               \", \".join(known_headers)))\n\u001b[0m\u001b[1;32m    117\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    118\u001b[0m         \u001b[0;31m# find the clustal version in the header line\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: ######################################## is not a known CLUSTAL header: CLUSTAL, PROBCONS, MUSCLE, MSAPROBS, Kalign"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO, AlignIO\n",
    "with open(\"../data/scrap/emboss_water-R20200322-004539-0954-87715542-p2m.aln.txt\", \"rU\") as handle:\n",
    "    for record in AlignIO.parse(handle, \"clustal\") :\n",
    "        print(record.id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-03-21T21:50:11.629634Z",
     "start_time": "2020-03-21T21:50:11.623933Z"
    }
   },
   "outputs": [],
   "source": [
    "RAW = \"../data/raw/\"\n",
    "outdir = \"../data/processed\"\n",
    "if not os.path.exists(outdir):\n",
    "    os.mkdir(outdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-03-21T21:50:11.644653Z",
     "start_time": "2020-03-21T21:50:11.631730Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['../data/raw/gisaid_cov2020_sequences.fasta',\n",
       " '../data/raw/NC_045512.2.fasta',\n",
       " '../data/raw/YP_009724390.1.fasta',\n",
       " '../data/raw/codons.txt',\n",
       " '../data/raw/GCF_009858895.2_ASM985889v3_genomic.gff']"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "glob.glob(RAW+\"/*\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-03-21T21:50:11.670834Z",
     "start_time": "2020-03-21T21:50:11.646403Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29903\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "SeqRecord(seq=Seq('ATTAAAGGTTTATACCTTCCCAGGTAACAAACCAACCAACTTTCGATCTCTTGT...AAA', SingleLetterAlphabet()), id='NC_045512.2', name='NC_045512.2', description='NC_045512.2 Severe acute respiratory syndrome coronavirus 2 isolate Wuhan-Hu-1, complete genome', dbxrefs=[])"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ref_fasta = SeqIO.read(os.path.join(RAW,'NC_045512.2.fasta'), format=\"fasta\")\n",
    "print(len(ref_fasta))\n",
    "ref_fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-03-21T21:50:11.701426Z",
     "start_time": "2020-03-21T21:50:11.672371Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>region</td>\n",
       "      <td>1</td>\n",
       "      <td>29903</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=NC_045512.2:1..29903;Dbxref=taxon:2697049;c...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>1</td>\n",
       "      <td>265</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id-NC_045512.2:1..265;gbkey=5'UTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>gene</td>\n",
       "      <td>266</td>\n",
       "      <td>21555</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene-GU280_gp01;Dbxref=GeneID:43740578;Name...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>266</td>\n",
       "      <td>13468</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-YP_009724389.1;Parent=gene-GU280_gp01;D...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>13468</td>\n",
       "      <td>21555</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-YP_009724389.1;Parent=gene-GU280_gp01;D...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>266</td>\n",
       "      <td>13483</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-YP_009725295.1;Parent=gene-GU280_gp01;D...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>gene</td>\n",
       "      <td>21563</td>\n",
       "      <td>25384</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene-GU280_gp02;Dbxref=GeneID:43740568;Name...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>21563</td>\n",
       "      <td>25384</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-YP_009724390.1;Parent=gene-GU280_gp02;D...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>gene</td>\n",
       "      <td>25393</td>\n",
       "      <td>26220</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene-GU280_gp03;Dbxref=GeneID:43740569;Name...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>25393</td>\n",
       "      <td>26220</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-YP_009724391.1;Parent=gene-GU280_gp03;D...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>gene</td>\n",
       "      <td>26245</td>\n",
       "      <td>26472</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene-GU280_gp04;Dbxref=GeneID:43740570;Name...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>26245</td>\n",
       "      <td>26472</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-YP_009724392.1;Parent=gene-GU280_gp04;D...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>gene</td>\n",
       "      <td>26523</td>\n",
       "      <td>27191</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene-GU280_gp05;Dbxref=GeneID:43740571;Name...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>26523</td>\n",
       "      <td>27191</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-YP_009724393.1;Parent=gene-GU280_gp05;D...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>gene</td>\n",
       "      <td>27202</td>\n",
       "      <td>27387</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene-GU280_gp06;Dbxref=GeneID:43740572;Name...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>27202</td>\n",
       "      <td>27387</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-YP_009724394.1;Parent=gene-GU280_gp06;D...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>gene</td>\n",
       "      <td>27394</td>\n",
       "      <td>27759</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene-GU280_gp07;Dbxref=GeneID:43740573;Name...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>27394</td>\n",
       "      <td>27759</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-YP_009724395.1;Parent=gene-GU280_gp07;D...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>gene</td>\n",
       "      <td>27756</td>\n",
       "      <td>27887</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene-GU280_gp08;Dbxref=GeneID:43740574;Name...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>27756</td>\n",
       "      <td>27887</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-YP_009725296.1;Parent=gene-GU280_gp08;D...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>gene</td>\n",
       "      <td>27894</td>\n",
       "      <td>28259</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene-GU280_gp09;Dbxref=GeneID:43740577;Name...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>27894</td>\n",
       "      <td>28259</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-YP_009724396.1;Parent=gene-GU280_gp09;D...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>gene</td>\n",
       "      <td>28274</td>\n",
       "      <td>29533</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene-GU280_gp10;Dbxref=GeneID:43740575;Name...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>28274</td>\n",
       "      <td>29533</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-YP_009724397.2;Parent=gene-GU280_gp10;D...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>gene</td>\n",
       "      <td>29558</td>\n",
       "      <td>29674</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene-GU280_gp11;Dbxref=GeneID:43740576;Name...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>29558</td>\n",
       "      <td>29674</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-YP_009725255.1;Parent=gene-GU280_gp11;D...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>29675</td>\n",
       "      <td>29903</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id-NC_045512.2:29675..29903;gbkey=3'UTR</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              0       1                2      3      4  5  6  7  \\\n",
       "0   NC_045512.2  RefSeq           region      1  29903  .  +  .   \n",
       "1   NC_045512.2  RefSeq   five_prime_UTR      1    265  .  +  .   \n",
       "2   NC_045512.2  RefSeq             gene    266  21555  .  +  .   \n",
       "3   NC_045512.2  RefSeq              CDS    266  13468  .  +  0   \n",
       "4   NC_045512.2  RefSeq              CDS  13468  21555  .  +  0   \n",
       "5   NC_045512.2  RefSeq              CDS    266  13483  .  +  0   \n",
       "6   NC_045512.2  RefSeq             gene  21563  25384  .  +  .   \n",
       "7   NC_045512.2  RefSeq              CDS  21563  25384  .  +  0   \n",
       "8   NC_045512.2  RefSeq             gene  25393  26220  .  +  .   \n",
       "9   NC_045512.2  RefSeq              CDS  25393  26220  .  +  0   \n",
       "10  NC_045512.2  RefSeq             gene  26245  26472  .  +  .   \n",
       "11  NC_045512.2  RefSeq              CDS  26245  26472  .  +  0   \n",
       "12  NC_045512.2  RefSeq             gene  26523  27191  .  +  .   \n",
       "13  NC_045512.2  RefSeq              CDS  26523  27191  .  +  0   \n",
       "14  NC_045512.2  RefSeq             gene  27202  27387  .  +  .   \n",
       "15  NC_045512.2  RefSeq              CDS  27202  27387  .  +  0   \n",
       "16  NC_045512.2  RefSeq             gene  27394  27759  .  +  .   \n",
       "17  NC_045512.2  RefSeq              CDS  27394  27759  .  +  0   \n",
       "18  NC_045512.2  RefSeq             gene  27756  27887  .  +  .   \n",
       "19  NC_045512.2  RefSeq              CDS  27756  27887  .  +  0   \n",
       "20  NC_045512.2  RefSeq             gene  27894  28259  .  +  .   \n",
       "21  NC_045512.2  RefSeq              CDS  27894  28259  .  +  0   \n",
       "22  NC_045512.2  RefSeq             gene  28274  29533  .  +  .   \n",
       "23  NC_045512.2  RefSeq              CDS  28274  29533  .  +  0   \n",
       "24  NC_045512.2  RefSeq             gene  29558  29674  .  +  .   \n",
       "25  NC_045512.2  RefSeq              CDS  29558  29674  .  +  0   \n",
       "26  NC_045512.2  RefSeq  three_prime_UTR  29675  29903  .  +  .   \n",
       "\n",
       "                                                    8  \n",
       "0   ID=NC_045512.2:1..29903;Dbxref=taxon:2697049;c...  \n",
       "1                ID=id-NC_045512.2:1..265;gbkey=5'UTR  \n",
       "2   ID=gene-GU280_gp01;Dbxref=GeneID:43740578;Name...  \n",
       "3   ID=cds-YP_009724389.1;Parent=gene-GU280_gp01;D...  \n",
       "4   ID=cds-YP_009724389.1;Parent=gene-GU280_gp01;D...  \n",
       "5   ID=cds-YP_009725295.1;Parent=gene-GU280_gp01;D...  \n",
       "6   ID=gene-GU280_gp02;Dbxref=GeneID:43740568;Name...  \n",
       "7   ID=cds-YP_009724390.1;Parent=gene-GU280_gp02;D...  \n",
       "8   ID=gene-GU280_gp03;Dbxref=GeneID:43740569;Name...  \n",
       "9   ID=cds-YP_009724391.1;Parent=gene-GU280_gp03;D...  \n",
       "10  ID=gene-GU280_gp04;Dbxref=GeneID:43740570;Name...  \n",
       "11  ID=cds-YP_009724392.1;Parent=gene-GU280_gp04;D...  \n",
       "12  ID=gene-GU280_gp05;Dbxref=GeneID:43740571;Name...  \n",
       "13  ID=cds-YP_009724393.1;Parent=gene-GU280_gp05;D...  \n",
       "14  ID=gene-GU280_gp06;Dbxref=GeneID:43740572;Name...  \n",
       "15  ID=cds-YP_009724394.1;Parent=gene-GU280_gp06;D...  \n",
       "16  ID=gene-GU280_gp07;Dbxref=GeneID:43740573;Name...  \n",
       "17  ID=cds-YP_009724395.1;Parent=gene-GU280_gp07;D...  \n",
       "18  ID=gene-GU280_gp08;Dbxref=GeneID:43740574;Name...  \n",
       "19  ID=cds-YP_009725296.1;Parent=gene-GU280_gp08;D...  \n",
       "20  ID=gene-GU280_gp09;Dbxref=GeneID:43740577;Name...  \n",
       "21  ID=cds-YP_009724396.1;Parent=gene-GU280_gp09;D...  \n",
       "22  ID=gene-GU280_gp10;Dbxref=GeneID:43740575;Name...  \n",
       "23  ID=cds-YP_009724397.2;Parent=gene-GU280_gp10;D...  \n",
       "24  ID=gene-GU280_gp11;Dbxref=GeneID:43740576;Name...  \n",
       "25  ID=cds-YP_009725255.1;Parent=gene-GU280_gp11;D...  \n",
       "26         ID=id-NC_045512.2:29675..29903;gbkey=3'UTR  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "anno = pd.read_csv(glob.glob(RAW+\"/*.gff\")[0],sep=\"\\t\", comment=\"#\", header=None)\n",
    "anno"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-03-21T21:50:11.708522Z",
     "start_time": "2020-03-21T21:50:11.703041Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSSVLHSTQDLFLPFFSNVTWFHAIHVSGTNGTKRFDNPVLPFNDGVYFASTEKSNIIRGWIFGTTLDSKTQSLLIVNNATNVVIKVCEFQFCNDPFLGVYYHKNNKSWMESEFRVYSSANNCTFEYVSQPFLMDLEGKQGNFKNLREFVFKNIDGYFKIYSKHTPINLVRDLPQGFSALEPLVDLPIGINITRFQTLLALHRSYLTPGDSSSGWTAGAAAYYVGYLQPRTFLLKYNENGTITDAVDCALDPLSETKCTLKSFTVEKGIYQTSNFRVQPTESIVRFPNITNLCPFGEVFNATRFASVYAWNRKRISNCVADYSVLYNSASFSTFKCYGVSPTKLNDLCFTNVYADSFVIRGDEVRQIAPGQTGKIADYNYKLPDDFTGCVIAWNSNNLDSKVGGNYNYLYRLFRKSNLKPFERDISTEIYQAGSTPCNGVEGFNCYFPLQSYGFQPTNGVGYQPYRVVVLSFELLHAPATVCGPKKSTNLVKNKCVNFNFNGLTGTGVLTESNKKFLPFQQFGRDIADTTDAVRDPQTLEILDITPCSFGGVSVITPGTNTSNQVAVLYQDVNCTEVPVAIHADQLTPTWRVYSTGSNVFQTRAGCLIGAEHVNNSYECDIPIGAGICASYQTQTNSPRRARSVASQSIIAYTMSLGAENSVAYSNNSIAIPTNFTISVTTEILPVSMTKTSVDCTMYICGDSTECSNLLLQYGSFCTQLNRALTGIAVEQDKNTQEVFAQVKQIYKTPPIKDFGGFNFSQILPDPSKPSKRSFIEDLLFNKVTLADAGFIKQYGDCLGDIAARDLICAQKFNGLTVLPPLLTDEMIAQYTSALLAGTITSGWTFGAGAALQIPFAMQMAYRFNGIGVTQNVLYENQKLIANQFNSAIGKIQDSLSSTASALGKLQDVVNQNAQALNTLVKQLSSNFGAISSVLNDILSRLDKVEAEVQIDRLITGRLQSLQTYVTQQLIRAAEIRASANLAATKMSECVLGQSKRVDFCGKGYHLMSFPQSAPHGVVFLHVTYVPAQEKNFTTAPAICHDGKAHFPREGVFVSNGTHWFVTQRNFYEPQIITTDNTFVSGNCDVVIGIVNNTVYDPLQPELDSFKEELDKYFKNHTSPDVDLGDISGINASVVNIQKEIDRLNEVAKNLNESLIDLQELGKYEQYIKWPWYIWLGFIAGLIAIVMVTIMLCCMTSCCSCLKGCCSCGSCCKFDEDDSEPVLKGVKLHYT'"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "S_protein_name = \"YP_009724390.1\"\n",
    "S_protein_fasta = SeqIO.read(os.path.join(RAW, S_protein_name+\".fasta\"),format='fasta')\n",
    "S_protein_fasta = str(S_protein_fasta.seq)\n",
    "S_protein_fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-03-21T21:50:11.918035Z",
     "start_time": "2020-03-21T21:50:11.710029Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "322\n",
      "29854\n",
      "29857\n",
      "29845\n",
      "29845\n",
      "29845\n",
      "29845\n",
      "29845\n",
      "29845\n",
      "29899\n",
      "29848\n",
      "29839\n",
      "29859\n",
      "29855\n",
      "29882\n",
      "29859\n",
      "29891\n",
      "29838\n",
      "29882\n",
      "29882\n",
      "29883\n",
      "29882\n",
      "29808\n",
      "29839\n",
      "29839\n",
      "29839\n",
      "29803\n",
      "29839\n",
      "29891\n",
      "29891\n",
      "29898\n",
      "29809\n",
      "29853\n",
      "29782\n",
      "29881\n",
      "29881\n",
      "29868\n",
      "29872\n",
      "322\n",
      "29833\n",
      "29845\n",
      "29882\n",
      "29836\n",
      "29782\n",
      "29882\n",
      "29843\n",
      "29844\n",
      "29882\n",
      "29882\n",
      "395\n",
      "29882\n",
      "29863\n",
      "29891\n",
      "29891\n",
      "29891\n",
      "29891\n",
      "29891\n",
      "29890\n",
      "29473\n",
      "29891\n",
      "29891\n",
      "29878\n",
      "29878\n",
      "29878\n",
      "29878\n",
      "29848\n",
      "29889\n",
      "29890\n",
      "29883\n",
      "29879\n",
      "29898\n",
      "29878\n",
      "29741\n",
      "29782\n",
      "29882\n",
      "29858\n",
      "29882\n",
      "29891\n",
      "29891\n",
      "182\n",
      "357\n",
      "357\n",
      "357\n",
      "29768\n",
      "29811\n",
      "290\n",
      "290\n",
      "290\n",
      "29878\n",
      "29878\n",
      "29409\n",
      "29829\n",
      "29843\n",
      "29882\n",
      "29801\n",
      "29806\n",
      "29806\n",
      "29806\n",
      "29795\n",
      "29865\n",
      "29865\n",
      "29891\n",
      "29889\n",
      "29842\n",
      "29864\n",
      "29758\n",
      "29891\n",
      "29845\n",
      "29867\n",
      "29785\n",
      "29874\n",
      "29903\n",
      "29871\n",
      "29891\n",
      "29873\n",
      "29872\n",
      "29878\n",
      "29873\n",
      "29862\n",
      "29891\n",
      "29890\n",
      "29818\n",
      "29891\n",
      "29891\n",
      "29891\n",
      "29903\n",
      "29882\n",
      "29882\n",
      "29882\n",
      "29886\n",
      "29911\n",
      "1260\n",
      "1260\n",
      "1260\n",
      "1260\n",
      "1260\n",
      "1260\n",
      "1260\n",
      "1260\n",
      "1260\n",
      "1260\n",
      "29870\n",
      "1260\n",
      "29851\n",
      "29843\n",
      "294\n",
      "294\n",
      "294\n",
      "294\n",
      "294\n",
      "294\n",
      "287\n",
      "287\n",
      "107\n",
      "107\n",
      "107\n",
      "107\n",
      "287\n",
      "287\n",
      "287\n",
      "107\n",
      "107\n",
      "107\n",
      "29889\n",
      "321\n",
      "321\n",
      "321\n",
      "321\n",
      "321\n",
      "321\n",
      "321\n",
      "321\n",
      "321\n",
      "321\n",
      "29848\n",
      "27213\n",
      "29882\n",
      "3822\n",
      "3822\n",
      "3822\n",
      "3822\n",
      "3822\n",
      "3822\n",
      "3822\n",
      "3822\n",
      "3822\n",
      "3822\n",
      "3822\n",
      "3822\n",
      "3822\n",
      "3822\n",
      "3822\n",
      "3822\n",
      "3822\n",
      "29848\n",
      "29848\n",
      "29201\n",
      "29756\n",
      "29870\n",
      "29891\n",
      "29927\n",
      "29902\n",
      "29902\n",
      "29806\n",
      "29903\n",
      "29903\n",
      "29865\n",
      "29903\n",
      "29782\n",
      "29846\n",
      "29848\n",
      "29845\n",
      "29843\n",
      "29838\n",
      "29697\n",
      "29851\n",
      "29903\n",
      "29858\n",
      "29879\n",
      "29864\n",
      "29875\n",
      "29845\n",
      "29788\n",
      "29871\n",
      "29782\n",
      "29782\n",
      "29903\n",
      "29879\n",
      "29635\n",
      "29860\n",
      "29903\n",
      "29887\n",
      "29893\n",
      "3822\n",
      "3822\n",
      "3822\n",
      "3822\n",
      "29851\n",
      "29870\n",
      "29866\n",
      "3822\n",
      "3822\n",
      "3822\n",
      "3822\n",
      "29854\n",
      "29851\n",
      "322\n",
      "363\n",
      "363\n",
      "29782\n",
      "29782\n",
      "29903\n",
      "29903\n",
      "29903\n",
      "29903\n",
      "29903\n",
      "29903\n",
      "29903\n",
      "29903\n",
      "29858\n",
      "29858\n",
      "29858\n",
      "29858\n",
      "29786\n",
      "29786\n",
      "29783\n",
      "29783\n",
      "29783\n",
      "29786\n",
      "29786\n",
      "29786\n",
      "29783\n",
      "29786\n",
      "29783\n",
      "29783\n",
      "29783\n",
      "29786\n",
      "29903\n",
      "29409\n",
      "29409\n",
      "29409\n",
      "29409\n",
      "29409\n",
      "29409\n",
      "29409\n",
      "29807\n",
      "29835\n",
      "29808\n",
      "29809\n",
      "29882\n",
      "29882\n",
      "29882\n",
      "29882\n",
      "29882\n",
      "29882\n",
      "29882\n",
      "29882\n",
      "29882\n",
      "29882\n",
      "29882\n",
      "29882\n",
      "29882\n",
      "29882\n",
      "29882\n",
      "29882\n",
      "29882\n",
      "29786\n",
      "29867\n",
      "29903\n",
      "29903\n",
      "29903\n",
      "29903\n",
      "29903\n",
      "29903\n",
      "363\n",
      "363\n",
      "363\n",
      "363\n",
      "29890\n",
      "29887\n",
      "29903\n",
      "29891\n",
      "29870\n",
      "29890\n",
      "29865\n",
      "29871\n",
      "29903\n",
      "29903\n",
      "29903\n",
      "29903\n",
      "29903\n",
      "29799\n",
      "29903\n",
      "29903\n",
      "29853\n",
      "29903\n",
      "29903\n",
      "29676\n",
      "29833\n",
      "29872\n",
      "29856\n",
      "29860\n",
      "29850\n",
      "29867\n",
      "29852\n",
      "29861\n",
      "29879\n",
      "29851\n",
      "29862\n",
      "29882\n",
      "29850\n",
      "29873\n",
      "29852\n",
      "29859\n",
      "29856\n",
      "28757\n",
      "26840\n",
      "28503\n",
      "29358\n",
      "29903\n",
      "29903\n",
      "29903\n",
      "29903\n",
      "29903\n",
      "29903\n",
      "29903\n",
      "29903\n",
      "29903\n",
      "29903\n",
      "29903\n",
      "29903\n",
      "29903\n",
      "29903\n",
      "29903\n",
      "29903\n",
      "29903\n",
      "29903\n",
      "29903\n",
      "29903\n",
      "29903\n",
      "29903\n",
      "29903\n",
      "29903\n",
      "29903\n",
      "29903\n",
      "29903\n",
      "29783\n",
      "29783\n",
      "29783\n",
      "29783\n",
      "29786\n",
      "29786\n",
      "29786\n",
      "29783\n",
      "29786\n",
      "29783\n",
      "29786\n",
      "29786\n",
      "29783\n",
      "29786\n",
      "29759\n",
      "25342\n",
      "28557\n",
      "29900\n",
      "29900\n",
      "29900\n",
      "29900\n",
      "29900\n",
      "29870\n",
      "402\n",
      "402\n",
      "400\n",
      "29903\n",
      "387\n",
      "315\n",
      "29890\n",
      "1065\n",
      "29893\n",
      "29892\n",
      "29891\n",
      "29899\n",
      "29896\n",
      "29891\n",
      "29891\n",
      "29852\n",
      "29903\n",
      "369\n",
      "29848\n",
      "29859\n",
      "29825\n",
      "402\n",
      "29463\n",
      "29462\n",
      "29463\n",
      "29786\n",
      "29775\n",
      "29786\n",
      "29786\n",
      "29780\n",
      "29786\n",
      "29786\n",
      "29786\n",
      "29786\n",
      "29786\n",
      "29783\n",
      "29786\n",
      "29783\n",
      "29786\n",
      "29783\n",
      "29786\n",
      "29786\n",
      "29786\n",
      "29786\n",
      "29786\n",
      "29783\n",
      "29786\n",
      "29786\n",
      "29786\n",
      "29786\n",
      "29786\n",
      "29786\n",
      "29786\n",
      "29786\n",
      "29786\n",
      "29786\n",
      "29786\n",
      "29786\n",
      "29786\n",
      "29786\n",
      "29786\n",
      "29786\n",
      "29786\n",
      "29783\n",
      "29783\n",
      "29786\n",
      "29783\n",
      "29786\n",
      "29786\n",
      "29786\n",
      "29786\n",
      "29786\n",
      "29786\n",
      "29864\n",
      "402\n",
      "29903\n",
      "29622\n",
      "29882\n",
      "29880\n",
      "29882\n",
      "29882\n",
      "29882\n",
      "29882\n",
      "29882\n",
      "29897\n",
      "29903\n",
      "29903\n",
      "29903\n",
      "29732\n",
      "29325\n",
      "29897\n",
      "29563\n",
      "29579\n",
      "29903\n",
      "29903\n",
      "29750\n",
      "29903\n",
      "29739\n",
      "29897\n",
      "29903\n",
      "29890\n",
      "29890\n",
      "29745\n",
      "29632\n",
      "29572\n",
      "29843\n",
      "29843\n",
      "29903\n",
      "29903\n",
      "29903\n",
      "29903\n",
      "29903\n",
      "29903\n",
      "29849\n",
      "29611\n",
      "29611\n",
      "29782\n",
      "29782\n",
      "29782\n",
      "29852\n",
      "29903\n",
      "29793\n",
      "29945\n",
      "29903\n",
      "253\n",
      "322\n",
      "322\n",
      "399\n",
      "87\n",
      "29764\n",
      "29764\n",
      "29789\n",
      "29789\n",
      "29789\n",
      "29789\n",
      "29789\n",
      "29903\n",
      "29903\n",
      "29903\n",
      "29903\n",
      "29903\n",
      "29775\n",
      "29775\n",
      "29775\n",
      "29775\n",
      "29772\n",
      "29775\n",
      "29775\n",
      "29875\n",
      "29775\n",
      "29775\n",
      "29775\n",
      "29874\n",
      "29775\n",
      "29775\n",
      "29772\n",
      "29775\n",
      "29775\n",
      "29772\n",
      "29772\n",
      "29775\n",
      "29786\n",
      "29772\n",
      "29775\n",
      "29775\n",
      "29772\n",
      "29772\n",
      "29772\n",
      "29751\n",
      "29772\n",
      "29772\n",
      "29772\n",
      "29772\n",
      "29772\n",
      "29772\n",
      "29775\n",
      "29775\n",
      "29775\n",
      "29775\n",
      "29775\n",
      "29775\n",
      "29764\n",
      "29764\n",
      "29764\n",
      "29764\n",
      "587\n",
      "595\n",
      "587\n",
      "29850\n",
      "29796\n",
      "29835\n",
      "29845\n",
      "29783\n",
      "29782\n",
      "29782\n",
      "29782\n",
      "29782\n",
      "29903\n",
      "29903\n",
      "29903\n",
      "29903\n",
      "29903\n",
      "29903\n",
      "29903\n",
      "29903\n",
      "29903\n",
      "29903\n",
      "29789\n",
      "29790\n",
      "29755\n",
      "29861\n",
      "29870\n",
      "29745\n",
      "29575\n",
      "29898\n",
      "29704\n",
      "363\n",
      "406\n",
      "29862\n",
      "29860\n",
      "29825\n",
      "29782\n",
      "29302\n",
      "29723\n",
      "29697\n",
      "29727\n",
      "29903\n",
      "29902\n",
      "29902\n",
      "29897\n",
      "29903\n",
      "29903\n",
      "29903\n",
      "29902\n",
      "29903\n",
      "29903\n",
      "29902\n",
      "29903\n",
      "27871\n",
      "29768\n",
      "29807\n",
      "29809\n",
      "29809\n",
      "29807\n",
      "29806\n",
      "29809\n",
      "29806\n",
      "524\n",
      "29903\n",
      "29855\n",
      "29903\n",
      "29633\n",
      "29644\n",
      "29896\n",
      "29903\n"
     ]
    }
   ],
   "source": [
    "# Load genomic sequences\n",
    "## Drop sequences with less than 27kb\n",
    "all_genomic_sequences = list(SeqIO.parse(os.path.join(RAW,\"gisaid_cov2020_sequences.fasta\"), \"fasta\"))\n",
    "for seq in all_genomic_sequences:\n",
    "    print(len(seq.seq))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Preprocess\n",
    "\n",
    "### a. Filter annotation for the S-protein.\n",
    "Confirm the protein sequence aligns with the genome annotation.\n",
    "- Extract the genomic region from annotation and take the sequence\n",
    "- Translate sequence\n",
    "- See if aligns with the Protein fasta sequence\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-03-20T21:29:10.486318Z",
     "start_time": "2020-03-20T21:29:10.479008Z"
    }
   },
   "source": [
    "See below that there are no introns,  so we can just use that sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-03-21T21:50:11.932859Z",
     "start_time": "2020-03-21T21:50:11.919692Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>21563</td>\n",
       "      <td>25384</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-YP_009724390.1;Parent=gene-GU280_gp02;D...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             0       1    2      3      4  5  6  7  \\\n",
       "7  NC_045512.2  RefSeq  CDS  21563  25384  .  +  0   \n",
       "\n",
       "                                                   8  \n",
       "7  ID=cds-YP_009724390.1;Parent=gene-GU280_gp02;D...  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "anno_S = anno[anno[8].str.contains(S_protein_name)]\n",
    "anno_S"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Extract protein sequence from this region"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-03-21T21:50:11.942964Z",
     "start_time": "2020-03-21T21:50:11.935425Z"
    }
   },
   "outputs": [],
   "source": [
    "S_gene_seq = ref_fasta.seq[anno_S[3].values[0]-1: anno_S[4].values[0]]\n",
    "annotation_protein_sequence = str(Seq.translate(S_gene_seq))\n",
    "annotation_protein_sequence = annotation_protein_sequence.replace(\"*\",\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-03-21T21:50:11.957912Z",
     "start_time": "2020-03-21T21:50:11.944756Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "S_gene_seq = SeqRecord.SeqRecord(S_gene_seq, id=S_protein_name)\n",
    "SeqIO.write(S_gene_seq, open(\"../data/S_gene.fasta\",'w'),format='fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-03-21T21:50:11.970550Z",
     "start_time": "2020-03-21T21:50:11.959551Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "1273\n",
      "1273\n"
     ]
    }
   ],
   "source": [
    "print((annotation_protein_sequence) == (S_protein_fasta))\n",
    "print(len(S_protein_fasta))\n",
    "print(len(annotation_protein_sequence))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-03-21T21:50:11.982627Z",
     "start_time": "2020-03-21T21:50:11.972919Z"
    },
    "code_folding": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annotation_protein_mismatch = []\n",
    "for ind, val in enumerate(annotation_protein_sequence):\n",
    "    if S_protein_fasta[ind] != val:\n",
    "        annotation_protein_mismatch.append((ind, val))\n",
    "annotation_protein_mismatch"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### b. remove any sequence less than 20kb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-03-21T21:50:11.992575Z",
     "start_time": "2020-03-21T21:50:11.984357Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of sequences beforeM filtering: 664\n",
      "Number of sequences after filtering: 563\n"
     ]
    }
   ],
   "source": [
    "print(f\"Number of sequences beforeM filtering: {len(all_genomic_sequences)}\")\n",
    "all_genomic_sequences_filt = []\n",
    "for seq in all_genomic_sequences:\n",
    "    if len(seq.seq) >= 20000:\n",
    "        all_genomic_sequences_filt.append(seq)\n",
    "print(f\"Number of sequences after filtering: {len(all_genomic_sequences_filt)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. Convert gene sequence to protein and save fasta file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5. Plot similarity"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 6. Run MSA on protein using clustalw"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-03-21T02:15:44.382209Z",
     "start_time": "2020-03-21T02:15:42.013Z"
    },
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "from coevolve.msa import clustalw_query"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from coevolve.config import CLUSTALW_DIR, CLUSTALW_CMD, NUM_CORES, CLUSTALW_CLEAN_DIR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "input_file = open(join(outdir,\"proteins.fasta\")\n",
    "clustalw_query.extract_msa(input_file, outdir, CLUSTALW_CMD, \"isshamie@ucsd.edu\",\n",
    "                to_overwrite=False, web=True,\n",
    "                local_backup=True, num_cores=NUM_CORES, verbose=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 7. Run EVCouplings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
